dawn/src/dawn/CodeGen/Cuda-ico/CudaIcoCodeGen.cpp

```diff
@@ -315,28 +315,28 @@ void CudaIcoCodeGen::generateRunFun(
         switch(loc) {
         case ast::LocationType::Cells:
           return (iterSpace.has_value() ? "mesh_.Domain({::dawn::LocationType::Cells," +
-                                              spaceMagicNumToEnum(iterSpace->upperBound()) + "," +
+                                              spaceMagicNumToEnum(iterSpace->upperLevel()) + "," +
                                               std::to_string(iterSpace->upperOffset()) + "})" +
                                               "- mesh_.Domain({::dawn::LocationType::Cells," +
-                                              spaceMagicNumToEnum(iterSpace->lowerBound()) + "," +
-                                              std::to_string(iterSpace->lowerOffset()) + "})"
+                                              spaceMagicNumToEnum(iterSpace->lowerLevel()) + "," +
+                                              std::to_string(iterSpace->lowerOffset()) + "}) + 1"
                                         : "mesh_.NumCells");
           break;
         case ast::LocationType::Edges:
           return (iterSpace.has_value() ? "mesh_.Domain({::dawn::LocationType::Edges," +
-                                              spaceMagicNumToEnum(iterSpace->upperBound()) + "," +
+                                              spaceMagicNumToEnum(iterSpace->upperLevel()) + "," +
                                               std::to_string(iterSpace->upperOffset()) + "})" +
                                               "- mesh_.Domain({::dawn::LocationType::Edges," +
-                                              spaceMagicNumToEnum(iterSpace->lowerBound()) + "," +
-                                              std::to_string(iterSpace->lowerOffset()) + "})"
+                                              spaceMagicNumToEnum(iterSpace->lowerLevel()) + "," +
+                                              std::to_string(iterSpace->lowerOffset()) + "}) + 1"
                                         : "mesh_.NumEdges");
         case ast::LocationType::Vertices:
           return (iterSpace.has_value() ? "mesh_.Domain({::dawn::LocationType::Vertices," +
-                                              spaceMagicNumToEnum(iterSpace->upperBound()) + "," +
+                                              spaceMagicNumToEnum(iterSpace->upperLevel()) + "," +
                                               std::to_string(iterSpace->upperOffset()) + "})" +
                                               "- mesh_.Domain({::dawn::LocationType::Vertices," +
-                                              spaceMagicNumToEnum(iterSpace->lowerBound()) + "," +
-                                              std::to_string(iterSpace->lowerOffset()) + "})"
+                                              spaceMagicNumToEnum(iterSpace->lowerLevel()) + "," +
+                                              std::to_string(iterSpace->lowerOffset()) + "}) + 1"
                                         : "mesh_.NumVertices");
```

Comment:
Noticed that we have a lot of code repetition here: can factorize out common part and leave switch to map location type enum to `Cells`/`Edges`/`Vertices` string (we should have this function somewhere, maybe to the singular version of the string?).
```
          return (iterSpace.has_value() ? "mesh_.Domain({::dawn::LocationType::" + toStringPlural(loc) + "," +
                                              spaceMagicNumToEnum(iterSpace->upperBound()) + "," +
                                              spaceMagicNumToEnum(iterSpace->upperLevel()) + "," +
                                              std::to_string(iterSpace->upperOffset()) + "})" +
                                              "- mesh_.Domain({::dawn::LocationType::" + toStringPlural(loc) + "," +
                                              spaceMagicNumToEnum(iterSpace->lowerBound()) + "," +
                                              std::to_string(iterSpace->lowerOffset()) + "})"
                                              spaceMagicNumToEnum(iterSpace->lowerLevel()) + "," +
                                              std::to_string(iterSpace->lowerOffset()) + "}) + 1"
                                        : "mesh_.Num" + toStringPlural(loc) + ");
```